core/src/test/java/org/apache/iceberg/catalog/CatalogTests.java

```diff
@@ -3167,6 +3170,687 @@ public void testRegisterExistingTable() {
+  }
+
+  @Test
+  public void testCreateTagWithInvalidSnapshotId() {
+    C catalog = catalog();
+
+    if (requiresNamespaceCreate()) {
+      catalog.createNamespace(NS);
+    }
+
+    Table table = catalog.buildTable(TABLE, SCHEMA).create();
+
+    String tagName = "corrupted-checkpoint";
+    long invalidSnapshotId = 999999L; // Non-existent snapshot ID
+
+    // Attempt to create tag with invalid snapshot ID should fail
+    assertThatThrownBy(() -> table.manageSnapshots().createTag(tagName, invalidSnapshotId).commit())
+        .isInstanceOf(ValidationException.class)
+        .hasMessageContaining("unknown snapshot");
+  }
+
+  @Test
+  public void testSetRetentionPolicyOnNonExistentBranch() {
+    C catalog = catalog();
+
+    if (requiresNamespaceCreate()) {
+      catalog.createNamespace(NS);
+    }
+
+    Table table = catalog.buildTable(TABLE, SCHEMA).create();
+
+    String nonExistentBranch = "missing-data-pipeline";
+
+    // Attempt to set retention policy on non-existent branch should fail
+    assertThatThrownBy(
+            () -> table.manageSnapshots().setMinSnapshotsToKeep(nonExistentBranch, 5).commit())
+        .isInstanceOf(IllegalArgumentException.class)
+        .hasMessageContaining("Branch does not exist: %s", nonExistentBranch);
+  }
+
+  @Test
+  public void testSetRetentionPolicyOnTag() {
+    C catalog = catalog();
+
+    if (requiresNamespaceCreate()) {
+      catalog.createNamespace(NS);
+    }
+
+    Table table = catalog.buildTable(TABLE, SCHEMA).create();
+    table.newFastAppend().appendFile(FILE_A).commit();
+    Snapshot currentSnapshot = table.currentSnapshot();
+
+    String tagName = "analytics-checkpoint";
+    table.manageSnapshots().createTag(tagName, currentSnapshot.snapshotId()).commit();
+
+    // Attempt to set branch-specific retention policy on tag should fail
+    assertThatThrownBy(() -> table.manageSnapshots().setMinSnapshotsToKeep(tagName, 5).commit())
+        .isInstanceOf(IllegalArgumentException.class)
+        .hasMessageContaining("Tags do not support setting minSnapshotsToKeep");
+  }
+
+  @Test
+  public void testReplaceBranchWithInvalidSnapshotId() {
+    C catalog = catalog();
+
+    if (requiresNamespaceCreate()) {
+      catalog.createNamespace(NS);
+    }
+
+    Table table = catalog.buildTable(TABLE, SCHEMA).create();
+    table.newFastAppend().appendFile(FILE_A).commit();
+
+    String branchName = "analytics-branch";
+    table.manageSnapshots().createBranch(branchName).commit();
+
+    long invalidSnapshotId = 999999L; // Non-existent snapshot ID
+
+    // Attempt to replace branch with invalid snapshot ID should fail
+    assertThatThrownBy(
+            () -> table.manageSnapshots().replaceBranch(branchName, invalidSnapshotId).commit())
+        .isInstanceOf(ValidationException.class)
+        .hasMessageContaining("unknown snapshot");
+  }
+
+  @Test
+  public void testReplaceTagWithInvalidSnapshotId() {
+    C catalog = catalog();
```

Comment:
Same as above, just validating failure on client, never hits the catalog. Should have coverage in TestSnapshotManager